### PR TITLE
(GSoC 2022, 8.2) TVB BIDS data builder and dir monitor module 

### PR DESCRIPTION
**Pull request by Spectre-ak**
Google Summer of Code 2022
Project: [Separate tool to upload data into TVB](https://neurostars.org/t/gsoc-2022-project-idea-8-2-separate-tool-to-upload-data-into-tvb-175-350-h/21392)
Mentors: @popaula937 @liadomide 

This PR contains a module which can build BIDS dataset and import them into TVB projects. Along with this, the module also provides helper class to monitor a BIDS directory and build datasets accordingly when new files are added.

BIDSDataBuilder

- Class for building BIDS dataset by providing 1) a set of json files and BIDS root directory or 2) a datatype from the subject directory (e.g. net, ts, coord, spatial)
- Contains utils for finding all dependencies of a subject's json file
- Produces a zip file
- Takes following args
   - BIDS_ROOT_DIR (reqd.)- BIDS root dir from which dataset will be created
   - BIDS_DATA_TO_IMPORT - Accepts BIDSImporter net, ts, coord, spatial token  to build subject type specific dataset For e.g. if BIDSImporter.TS_TOKEN is provided then, it'll build a dataset containing only TS data and their dependencies
   - INITIAL_JSON_FILES - A set of initial json files with which bids dataset to be built. This is used by BIDSDirWatcher

BIDSDirWatcher

- Class for monitoring a BIDS directory and also builds datasets accordingly on new files
- Runs two threads in parallel for 1) observing a directory and 2) building/importing dataset into TVB
- Importing thread runs on a fixed interval (configurable), which also acts as a buffer time when large files are added
- Contains utils for observing specific (subjects) directories changes only
- Takes following args
    - DIRECTORY_TO_WATCH (reqd.)- BIDS root dir on which is to be monitored
    - UPLOAD_TRIGGER_INTERVAL - Importer thread interval in seconds
    - IMPORT_DATA_IN_TVB - A flag for only creating dataset and storing and not importing into TVB projects
    - TVB_PROJECT_ID - ID of TVB project on which dataset is to be imported

[launch_bids_monitor.py](tvb_framework/tvb/interfaces/rest/bids_monitoring/launch_bids_monitor.py) contains sample code

Build BIDS dataset for time_series

```python
   from tvb.interfaces.rest.bids_monitoring.bids_data_builder import BIDSDataBuilder
   from tvb.adapters.uploaders.bids_importer import BIDSImporter

   bids_data_builder = BIDSDataBuilder(BIDSImporter.TS_TOKEN, BIDS_DIR)
   zip_file_location = bids_data_builder.create_dataset_subjects()
   print(zip_file_location)
```

Monitor a BIDS directory for new files

```python
   from tvb.interfaces.rest.bids_monitoring.bids_dir_monitor import BIDSDirWatcher
   
   bids_dir_watcher = BIDSDirWatcher(
        DIRECTORY_TO_WATCH=BIDS_DIR,
        UPLOAD_TRIGGER_INTERVAL=20,
        IMPORT_DATA_IN_TVB=True
   )
   bids_dir_watcher.init_watcher()
```

### CODE REVIEW

File: tvb_framework/tvb/interfaces/rest/bids_monitoring/bids_data_builder.py [hunk 1]
```diff
@@ -0,0 +1,227 @@
+import os
```
Comment (liadomide) [inline tvb_framework/tvb/interfaces/rest/bids_monitoring/bids_data_builder.py:1]: We can merge only if you agree to preserve our copyright header (like in other tvb-framework python files).

File: tvb_framework/setup_bids_monitor.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+import shutil
+
+import setuptools
+from setuptools.command.egg_info import manifest_maker
+
+manifest_maker.template = 'MANIFEST_bids_monitor.in'
+
+VERSION = "1.0"
+
+TVB_TEAM = "Lia Domide, Paula Prodan, Bogdan Valean, Robert Vincze"
```
Comment (liadomide) [inline tvb_framework/setup_bids_monitor.py:45]: Here i think that only Akash and Paula should be listed as authors.

File: tvb_framework/setup_bids_monitor.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+This is used to package the BIDS directory monitor separately.
+"""
+
+import os
+import shutil
+
+import setuptools
+from setuptools.command.egg_info import manifest_maker
+
+manifest_maker.template = 'MANIFEST_bids_monitor.in'
```
Comment (liadomide) [inline tvb_framework/setup_bids_monitor.py:41]: Is this Manifest file missing from the PR ?

File: tvb_framework/setup_bids_monitor.py [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+TVB_TEAM = "Lia Domide, Paula Prodan, Bogdan Valean, Robert Vincze"
+
+TVB_INSTALL_REQUIREMENTS = ["tvb-rest-client", "watchdog"]
+
+# Packaging bids-monitor
+with open(os.path.join(os.path.dirname(__file__), 'README_bids-monitor.rst')) as fd:
+    DESCRIPTION = fd.read()
+
+setuptools.setup(name="bids-monitor",
```
Comment (liadomide) [inline tvb_framework/setup_bids_monitor.py:53]: Would you agree to rename the package into something with a `tvb-` prefix ?
Maybe `tvb-bids-monitor` ?

File: tvb_framework/tvb/interfaces/rest/bids_monitoring/launch_bids_monitor.py [hunk 5]
```diff
@@ -0,0 +1,46 @@
+import sys
+from tvb.interfaces.rest.bids_monitoring.bids_data_builder import BIDSDataBuilder
+from tvb.interfaces.rest.bids_monitoring.bids_dir_monitor import BIDSDirWatcher
+from tvb.adapters.uploaders.bids_importer import BIDSImporter
+
+
+BIDS_UPLOAD_CONTENT = BIDSImporter.NET_TOKEN
+BIDS_DIR = "C:/Users/upadh/Documents/GitHub/tvb-root/tvb_framework/tvb/interfaces/rest/client/examples/BIDS_DEMO_DATSET - Copy"
```
Comment (liadomide) [inline tvb_framework/tvb/interfaces/rest/bids_monitoring/launch_bids_monitor.py:8]: This hard-coded path might be wrong to keep here ?

File: tvb_framework/setup_bids_monitor.py [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+                 version=VERSION,
+                 packages=setuptools.find_packages(
+                     exclude=['tvb.interfaces.web', 'tvb.interfaces.web.*', 'tvb.interfaces.command',
+                              'tvb.interfaces.command.*', 'tvb.tests', 'tvb.tests.*']),
+                 include_package_data=True,
+                 install_requires=TVB_INSTALL_REQUIREMENTS,
+                 extras_require={'postgres': ["psycopg2"],
```
Comment (liadomide) [inline tvb_framework/setup_bids_monitor.py:60]: Is `psycopg` really a dependency ?
`pytest` would be nice to be used

**Review by liadomide — COMMENTED** [review]
Could we have some unit-tests for the new code ?

File: tvb_framework/setup_bids_monitor.py (hunk 6 above)
Comment (Spectre-ak) [reply tvb_framework/setup_bids_monitor.py:60]: These will be automatically included by `tvb-rest-client` which is a dependency itself

Comment (Spectre-ak) [reply tvb_framework/setup_bids_monitor.py:60]: I'll remove it for now

File: tvb_framework/tvb/interfaces/rest/bids_monitoring/launch_bids_monitor.py (hunk 5 above)
Comment (Spectre-ak) [reply tvb_framework/tvb/interfaces/rest/bids_monitoring/launch_bids_monitor.py:8]: yes, removing it

File: tvb_framework/setup_bids_monitor.py (hunk 4 above)
Comment (Spectre-ak) [reply tvb_framework/setup_bids_monitor.py:53]: sure

File: tvb_framework/setup_bids_monitor.py (hunk 3 above)
Comment (Spectre-ak) [reply tvb_framework/setup_bids_monitor.py:41]: oh yes, i forget to add it

File: tvb_framework/setup_bids_monitor.py (hunk 2 above)
Comment (Spectre-ak) [reply tvb_framework/setup_bids_monitor.py:45]: sure

File: tvb_framework/tvb/interfaces/rest/bids_monitoring/bids_data_builder.py (hunk 1 above)
Comment (Spectre-ak) [reply tvb_framework/tvb/interfaces/rest/bids_monitoring/bids_data_builder.py:1]: Okay, i'll add the header

**Comment by Spectre-ak** [comment]
Hi, I've added MANIFEST_bids_monitor.in, README_bids_monitor.rst and updated the code as per the suggestions.
Can you please check again? 
Thanks

File: tvb_framework/README_bids_monitor.rst [hunk 7]
```diff
@@ -0,0 +1,67 @@
+BIDS data builder and directory monitor
```
Comment (popaula937) [inline tvb_framework/README_bids_monitor.rst:1]: This file is duplicated, I suggest we keep it only under the bids_monitoring folder. The setup_bids_monitor.py should include it from there.

File: tvb_framework/setup_bids_monitor.py [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+TVB_INSTALL_REQUIREMENTS = ["tvb-rest-client", "watchdog"]
+
+# Packaging bids-monitor
+with open(os.path.join(os.path.dirname(__file__), 'README_bids_monitor.rst')) as fd:
+    DESCRIPTION = fd.read()
+
+setuptools.setup(name="tvb-bids-monitor",
+                 version=VERSION,
+                 packages=setuptools.find_packages(
+                     exclude=['tvb.interfaces.web', 'tvb.interfaces.web.*', 'tvb.interfaces.command',
```
Comment (popaula937) [inline tvb_framework/setup_bids_monitor.py:56]: All packages excluded in MANIFEST, should also be added here.

File: tvb_framework/tvb/interfaces/rest/bids_monitoring/README.md [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+
+
+[launch_bids_monitor.py](tvb_framework/tvb/interfaces/rest/bids_monitoring/launch_bids_monitor.py) contains sample code and it also accepts command line arguments. To start monitoring a bids directory run below command
+
+```sh
+   $ python launch_bids_monitor.py --rest-url=http://localhost:9090 --bids-dir=user/doc/BIDS_SAMPLE
+```
+
+where  `-rest-url` is the url on which TVB rest server is running and `-bids-dir` is the BIDS root directory which is to be monitored for new files
+
```
Comment (popaula937) [inline tvb_framework/tvb/interfaces/rest/bids_monitoring/README.md:35]: Would be worth mentioning here that the user needs to add new files under the BIDS subject directories in order for this module to trigger an import into TVB.

File: tvb_framework/tvb/interfaces/rest/bids_monitoring/bids_data_builder.py [hunk 10]
```diff
@@ -0,0 +1,257 @@
...
+        for p in data_files:
+            import_dependencies_paths.add(p)
+        for p in import_dependencies_paths:
+            final_paths_set.add(p)
+
+        return final_paths_set
+
+    def subject_build_data_check(self):
+        if self.bids_root_dir is None:
+            print("BIDS root directory is empty")
```
Comment (popaula937) [inline tvb_framework/tvb/interfaces/rest/bids_monitoring/bids_data_builder.py:241]: The print statements should be removed

File: tvb_framework/tvb/interfaces/rest/bids_monitoring/README.md (hunk 9 above)
Comment (Spectre-ak) [reply tvb_framework/tvb/interfaces/rest/bids_monitoring/README.md:35]: sure

File: tvb_framework/setup_bids_monitor.py (hunk 8 above)
Comment (Spectre-ak) [reply tvb_framework/setup_bids_monitor.py:56]: okay

File: tvb_framework/MANIFEST_bids_monitor.in [hunk 11]
```diff
@@ -0,0 +1,9 @@
+include LICENSE
+include README_bids_monitor.rst
+include tvb/config/logger/*.conf
+recursive-include tvb *
+recursive-exclude tvb *.pyc
```
[the remark this answers is not in the document]
Comment (Spectre-ak) [reply tvb_framework/MANIFEST_bids_monitor.in:5]: yes, makes sense

File: tvb_framework/tvb/interfaces/rest/bids_monitoring/bids_data_builder.py (hunk 10 above)
Comment (Spectre-ak) [reply tvb_framework/tvb/interfaces/rest/bids_monitoring/bids_data_builder.py:241]: removed

**Review by popaula937 — COMMENTED** [review]
Thanks for the changes! This form looks decent to me.
It would be great if you could add some unit tests in the remaining time.

File: tvb_framework/setup_bids_monitor.py [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+                 include_package_data=True,
+                 install_requires=TVB_INSTALL_REQUIREMENTS,
+                 extras_require={'test': ["pytest", "pytest-benchmark"]},
+                 description='A helper package containing BIDS directory monitor module',
+                 long_description=DESCRIPTION,
+                 license="GPL-3.0-or-later",
+                 author=TVB_TEAM,
+                 author_email='tvb.admin@thevirtualbrain.org',
+                 url='http://www.thevirtualbrain.org',
+                 download_url='https://github.com/the-virtual-brain/tvb-framework',
```
Comment (liadomide) [inline tvb_framework/setup_bids_monitor.py:69]: Download URL should be `https://github.com/the-virtual-brain/tvb-root`.
(I fix it also in the other `setup.py` as I think this was copied from there - our bad).

Comment (Spectre-ak) [reply tvb_framework/setup_bids_monitor.py:69]: i'll update it

**Comment by Spectre-ak** [comment]
Hi @popaula937 @liadomide , I've added unit tests TestBIDSDataBuilder, TestBIDSDirWatcher please take a look
thanks

**Pull request merged**